### ISSUE TITLE
Custom Error Formatting

### ISSUE THREAD

**Issue by sh0rez** [google/go-jsonnet#453]
Hey!

I would like to implement a custom `jsonnet.ErrorFormatter`, mostly because I would like to do screen formatting a bit different:

- `RUNTIME` / `STATIC` / `INTERNAL ERROR` in red
- newline before stacktrace
- stacktrace in grey

These are visual preferences and probably not worth discussing here.

However, I figured this was much harder than expected. While the `ErrorFormatter` interface exists, it cannot effectively be implemented:

- `Format()` receives multiple kinds of errors, it needs to decide how to render them individually
- `termErrorFormatter` solves this using a type switch. **BUT**: `errors.StaticError` is `internal`, so I cannot implement that very type switch myself
- `termErrorFormatter` is private, so I cannot re-use code from it. This forces me to copy that code over to do slight modifications

The use of `internal` in this repository has historically made it hard to implement certain features in downstream projects (https://github.com/google/go-jsonnet/pull/395, https://github.com/google/go-jsonnet/pull/327).

Is there a strong reason to do this? Because I would really like to import, compose and hack with all code of this repository, without certain subsets being reserved for use in here exclusively :)

**Comment by sbarzowski** [comment]
> Is there a strong reason to do this? Because I would really like to import, compose and hack with all code of this repository, without certain subsets being reserved for use in here exclusively :)

Yes, there is. We don't want to break the users. The internal APIs often depend on implementation details and concepts which we want to be free to change at any time. Another reason is that it minimizes the amount of irrelevant stuff faced by the users.

If you see an API which you think should be useful for external users, please open an issue.

> * `RUNTIME` / `STATIC` / `INTERNAL ERROR` in red
> * newline before stacktrace
> * stacktrace in grey
> 
> These are visual preferences and probably not worth discussing here.

Maybe not in this issue, but I'm interested in the discussion about how the errors should be formatted. I think we can make the upstream "pretty"-mode format better.

>  `errors.StaticError` is `internal`, so I cannot implement that very type switch myself

This is an oversight. There's no good reason to keep it internal if RuntimeError is not internal. Feel free to send a PR to move `internal/errors` to `errors`. I can also do it myself later.

> `termErrorFormatter` is private, so I cannot re-use code from it. This forces me to copy that code over to do slight modifications

I don't remember why I made it private originally. It predates most of the refactoring. I think i didn't want to expose the 

> Format() receives multiple kinds of errors, it needs to decide how to render them individually

I'm not really happy with how error API looks like in general and I'm thinking about making some deeper changes. With Linter out of the way, I think some heavy changes around error handling will come relatively soon. Some things I want to achieve:
* Better support for multiple static errors.
* Find an elegant solution to RUNTIME vs STATIC for errors found in the analysis in imported dependencies (in a sense they are static, because they are found when the source is analyzed, but also they happen during execution and have a stack trace).
* Clean, well-documented API for external formatters.
* A clean recommended way for external tools to use errors (e.g. for linter).
* Get rid of SetColorFormatter – whether or not such thing is injected should be an implementation detail of the formatter.
* Clean way to switch error formatting in CLI. 
* JSON format for errors for consumption by tools.
* Maybe distinguish errors and warnings (e.g. from linter).

**Comment by sh0rez** [comment]
> Some things I want to achieve:

> - Better support for multiple static errors.
> - Find an elegant solution to RUNTIME vs STATIC for errors found in the analysis in imported dependencies (in a sense they are static, because they are found when the source is analyzed, but also they happen during execution and have a stack trace).
> - ...

This sounds super cool!! Especially replacing the interface with a documented, more sensible one sounds incredibly useful. This is not pressing on our end, we can happily wait until cleanup happened here :)

A suggestion for above list: If a nativeFunc raised an error, print which nativeFunc this was. Maybe also figure out the line in the source Go source code this func is defined at (in case that's possible)

> Feel free to send a PR to move internal/errors to errors.

Sure thing!

> Maybe not in this issue, but I'm interested in the discussion about how the errors should be formatted. I think we can make the upstream "pretty"-mode format better.

Oh yes!! This was linked in the source code: https://elm-lang.org/news/compiler-errors-for-humans and makes some good points.

Generally I experienced the errors of Jsonnet quite hard at times, especially because the stacktrace includes a lot of things that are internal and make it hard to trace where it originated from.  
Also many error items aren't quite actionable and require good knowledge of language internals
